Guide update - How to configure and rebuild software RAID (#8252)
ovh/docs: pages/bare_metal_cloud/dedicated_servers/raid_soft_bios/guide.en-gb.md

---
title: Managing and rebuilding software RAID on servers in BIOS boot mode
excerpt: Find out how to manage and rebuild software RAID after a disk replacement on your server in BIOS mode
updated: 2025-08-xx
---

## Objective

Redundant Array of Independent Disks (RAID) is a technology that mitigates data loss on a server by replicating data across two or more disks.

The default RAID level for OVHcloud server installations is RAID 1, which doubles the space taken up by your data, effectively halving the useable disk space.

> **This guide explains how to manage and rebuild software RAID after a disk replacement on your server in BIOS mode**

Reviewer: ```suggestion
**This guide explains how to manage and rebuild software RAID after a disk replacement on your server in legacy boot (BIOS) mode**
```
Internally, we refer to the boot mode as UEFI or legacy, not BIOS. I think BIOS could be misleading as the UEFI setup menu is still sometimes called "BIOS" or "BIOS setup".

md3 : active raid1 nvme1n1p3[1] nvme0n1p3[0]
      497875968 blocks super 1.2 [2/2] [UU]
      bitmap: 2/4 pages [8KB], 65536KB chunk

md2 : active raid1 nvme1n1p2[1] nvme0n1p2[0]
      1046528 blocks super 1.2 [2/2] [UU]

unused devices: <none>
```

This command shows us that we have two software RAID devices currently set up, with **md3** being the largest one. This array consists of two partitions, which are known as **nvme1n1p3** and **nvme0n1p3**. 

The [UU] means that all the disks are working normally. A `_` would indicate a failed disk.

If you have a server with SATA disks, you would get the following results:

```sh
cat /proc/mdstat
Personalities : [raid1] [linear] [multipath] [raid0] [raid6] [raid5] [raid4] [raid10]
md3 : active raid1 sda3[0] sdb3[1]
      3904786432 blocks super 1.2 [2/2] [UU]
      bitmap: 2/30 pages [8KB], 65536KB chunk

md2 : active raid1 sda2[0] sdb2[1]
      1046528 blocks super 1.2 [2/2] [UU]

unused devices: <none>
```

Although this command returns our RAID volumes, it doesn't tell us the size of the partitions themselves. We can find this information with the following command:

```sh

> sudo fdisk -l

Reviewer: Why no prompt here? Some other commands have `[user@server_ip ~]#`

Disk /dev/sdb: 1.8 TiB, 2000398934016 bytes, 3907029168 sectors
Disk model: HGST HUS724020AL
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: F92B6C5B-2518-4B2D-8FF9-A311DED5845F

Device          Start        End    Sectors   Size Type

> /dev/sdb1        2048       4095       2048     1M BIOS boot

Reviewer: After reading the whole guide, I think we should explain how to reinstall GRUB to ensure this partition does something. If you just copy the partition table but do not reinstall GRUB, it won't contain the bootloader.
In your case, the server still boots because you replaced one disk but if you replace both, it won't.

I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 2E1DCCBA-8808-4D2B-BA33-9FEC3B96ADA8

Device          Start        End    Sectors   Size Type
/dev/sda1        2048       4095       2048     1M BIOS boot
/dev/sda2        4096 1864177663 1864173568 888.9G Linux RAID
/dev/sda3  1864177664 1865226239    1048576   512M Linux filesystem
/dev/sda4  1865226240 3907024895 2041798656 973.6G Linux RAID
/dev/sda5  3907025072 3907029134       4063     2M Linux filesystem

Disk /dev/md4: 973.5 GiB, 1045265645568 bytes, 2041534464 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

Disk /dev/md2: 888.8 GiB, 954321600512 bytes, 1863909376 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

The `fdisk -l` command also allows you to identify your partition type. This is an important information when it comes to rebuilding your RAID in case of a disk failure.

For **GPT** partitions, line 6 will display: `Disklabel type: gpt`.

> For **MBR** partitions, line 6 will display: `Disklabel type: dos`.

Reviewer: I would not document MBR at all. We've entirely abandoned it for new installations. Only 3% of the installations done in 2024 were in this case and starting from July 2025 it's 0%.

> Still going by the results of `fdisk -l`, we can see that `/dev/md2` consists of 888.8GB and `/dev/md4` contains 973.5GB. If we were to run the mount command we can also find out the layout of the disk.

Reviewer: You should mention that it's only the case when booted to disk, not in rescue.
Also, mount's output is pretty verbose. I'd just keep lsblk's

cgroup on /sys/fs/cgroup/rdma type cgroup (rw,nosuid,nodev,noexec,relatime,rdma)
cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
debugfs on /sys/kernel/debug type debugfs (rw,relatime)
hugetlbfs on /dev/hugepages type hugetlbfs (rw,relatime,pagesize=2M)
mqueue on /dev/mqueue type mqueue (rw,relatime)
systemd-1 on /proc/sys/fs/binfmt_misc type autofs (rw,relatime,fd=45,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=10340)
/dev/md4 on /home type ext3 (rw,relatime)
tmpfs on /run/user/1000 type tmpfs (rw,nosuid,nodev,relatime,size=3266552k,mode=700,uid=1000,gid=1000)
```

Alternatively, the `lsblk` command offers a different view of the partitions:

```sh

> lsblk

Reviewer: Why no prompt here? Some other commands have `[user@server_ip ~]#`

NAME    MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda       8:0    0   1.8T  0 disk
├─sda1    8:1    0     1M  0 part
├─sda2    8:2    0 888.9G  0 part
│ └─md2   9:2    0 888.8G  0 raid1 /
├─sda3    8:3    0   512M  0 part  [SWAP]
├─sda4    8:4    0 973.6G  0 part
│ └─md4   9:4    0 973.5G  0 raid1 /home

> └─sda5    8:5    0     2M  0 part

Reviewer: 1. You could maybe explain what this partition is so that users aren't worried about us not copying it when the RAID is replaced? https://cloudinit.readthedocs.io/en/latest/reference/datasources/configdrive.html
2. Adding `lsblk -f` would be nice, it shows the partition LABELs which contain additional info such as the config-2 label here

sdb       8:16   0   1.8T  0 disk
├─sdb1    8:17   0     1M  0 part
├─sdb2    8:18   0 888.9G  0 part
│ └─md2   9:2    0 888.8G  0 raid1 /
├─sdb3    8:19   0   512M  0 part  [SWAP]
└─sdb4    8:20   0 973.6G  0 part
  └─md4   9:4    0 973.5G  0 raid1 /home
```

We take note of the devices, partitions and their mount points. From the above commands and results, we have:

- Two RAID arrays: `/dev/md2` and `/dev/md4`.
- Partitions part of the RAID: `/` and `/home`.
- Partitions not part of the RAID: [SWAP].

### System verification

To check whether a server runs on BIOS mode or BIOS mode, run the following command:

```sh

> [user@server_ip ~]# [ -d /sys/firmware/efi ] && echo BIOS || echo BIOS

Reviewer: That should be moved at the beginning of the doc when you mention legacy/UEFI.

```

### Simulating a disk failure

Now that we have all the necessary information, we can simulate a disk failure and proceed with the tests. In this example, we will fail the main disk `sda`.

The preferred way to do this is via the OVHcloud rescue mode environment.

> First reboot in rescue menu and log in with the credentials provided.

Reviewer: ```suggestion
First reboot the server in rescue mode and log in with the provided credentials.
```

To remove a disk from the RAID, the first step is to mark it as **Failed** and remove the partitions from their respective RAID arrays.

```sh
cat /proc/mdstat

Personalities : [raid1] [linear] [multipath] [raid0] [raid6] [raid5] [raid4] [raid10]
md2 : active raid1 sda2[1] sdb2[0]
      931954688 blocks super 1.2 [2/2] [UU]
      bitmap: 2/7 pages [8KB], 65536KB chunk

md4 : active raid1 sda4[0] sdb4[1]
      1020767232 blocks super 1.2 [2/2] [UU]
      bitmap: 0/8 pages [0KB], 65536KB chunk
      
unused devices: <none>
```
From the above output, sda consists of two partitions in RAID which are **sda2** and **sda4**. 

#### Removing the failed disk

First we mark the partitions **sda2** and **sda4** as failed. 

```sh

> sudo mdadm --manage /dev/md2 --fail /dev/sda2

Reviewer: Needs a prompt here too (or no prompt at all for all commands)

# mdadm: set /dev/sda2 faulty in /dev/md2
```

```sh
sudo mdadm --manage /dev/md4 --fail /dev/sda4

# mdadm: set /dev/sda4 faulty in /dev/md4
```

We have now simulated a failure of the RAID, when we run the `cat /proc/mdstat` command, we have the following output

```sh
cat /proc/mdstat 

Personalities : [raid1] [linear] [multipath] [raid0] [raid6] [raid5] [raid4] [raid10]
md2 : active raid1 sda2[1](F) sdb2[0]
      931954688 blocks super 1.2 [2/2] [_U]
      bitmap: 2/7 pages [8KB], 65536KB chunk

md4 : active raid1 sda4[0](F) sdb4[1]
      1020767232 blocks super 1.2 [2/2] [_U]
      bitmap: 0/8 pages [0KB], 65536KB chunk
unused devices: <none>
```

As we can see above, the [F] next to the partitions indicates that the disk has failed or is faulty.

> Next, we remove these partitions from the RAID arrays to completely remove the disk from RAID.

Reviewer: ```suggestion
Next, we remove these partitions from the RAID arrays.
```
The second part is a bit redundant IMO.

```sh
sudo mdadm --manage /dev/md2 --remove /dev/sda2

# mdadm: hot removed /dev/sda2 from /dev/md2
```

```sh
sudo mdadm --manage /dev/md3 --remove /dev/sda3

# mdadm: hot removed /dev/nvme0n1p3 from /dev/md3
```

To make sure that we get a disk that is similar to an empty disk, we use the following command. Replace **sda** with your own values:

```sh
sudo shred -s10M -n1 /dev/sda
```

> The disk now appears as a new one:

Reviewer: ```suggestion
The disk now appears as a new, empty drive:
```

```sh
lsblk 
NAME    MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda       8:0    0   1.8T  0 disk
sdb       8:16   0   1.8T  0 disk
├─sdb1    8:17   0     1M  0 part
├─sdb2    8:18   0 888.9G  0 part
│ └─md2   9:2    0 888.8G  0 raid1 /
├─sdb3    8:19   0   512M  0 part  [SWAP]
└─sdb4    8:20   0 973.6G  0 part
  └─md4   9:4    0 973.5G  0 raid1 /home
```

### Rebuilding the RAID

The following steps document the RAID rebuild in rescue mode.

> [!primary]
> This process might be different depending on the operating system you have installed on your server. We recommend that you consult the official documentation of your operating system to have access to the proper commands.
>

> [!warning]
>
> For most servers in software RAID, after a disk replacement, the server is able to reboot in normal mode (on the healthy disk) and the rebuild can be done in normal mode. However, if the server is not able to reboot in normal mode after a disk replacement, it will be rebooted in rescue mode to proceed with the RAID rebuild.
>

#### Rebuilding the RAID after the main disk is replaced

> Once the disk has been replaced, we need to copy the partition table from the healthy disk (in this example, sdb ) to the new one (sda).

Reviewer: ```suggestion
Once the disk has been replaced, we need to copy the partition table from the healthy disk (in this example, sdb) to the new one (sda).
```

> [!tabs]
> **For GPT partitions**
>>
>> ```sh
>> sudo sgdisk -R /dev/sdX /dev/sdX
>> ```
>>
>> The command should be in this format: `sgdisk -R /dev/newdisk /dev/healthydisk`
>>
>> Example:
>>
>> ```sh
>> sudo sgdisk -R /dev/sda /dev/sdb
>> ```
>>
>> Once this is done, the next step is to randomize the GUID of the new disk to prevent GUID conflicts with other disks:
>>
>> ```sh
>> sudo sgdisk -G /dev/sda
>> ```
>>
>> If you receive a message like this:
>> 
>> ```console
>> Warning: The kernel is still using the old partition table.
>> The new table will be used at the next reboot or after you
>> run partprobe(8) or kpartx(8)
>> The operation has completed successfully.
>> ```
>>

> >> You can simply run the `partprobe` command or reboot the server when the raid rebuilt is done.

Reviewer: I'd maybe say something like
```suggestion
>> You can simply run the `partprobe` command. If you still cannot see the newly-created partitions (e.g. with `lsblk`), you need to reboot the server before continuing.
```

Also can you please make sure all instances of RAID are capitalized?

      [============>........]  recovery = 64.8% (822969856/1020767232) finish=7.2min speed=401664K/sec
      bitmap: 0/8 pages [0KB], 65536KB chunk
unused devices: <none>
```

We can verify the RAID details with the following command:

```sh
mdadm --detail /dev/md4

/dev/md4:
        Version : 1.2
     Creation Time : Tue Jan 24 15:35:02 2023
        Raid Level : raid1
        Array Size : 1020767232 (973.48 GiB 1045.27 GB)
     Used Dev Size : 1020767232 (973.48 GiB 1045.27 GB)
      Raid Devices : 2
     Total Devices : 2
       Persistence : Superblock is persistent

     Intent Bitmap : Internal

       Update Time : Tue Jan 24 17:02:55 2023
             State : clean
    Active Devices : 2
   Working Devices : 2
    Failed Devices : 0
     Spare Devices : 0

 Rebuild Status : 21% complete

           UUID : 7f39d062:9f16a016:a4d2adc2:26fd5302
         Events : 0.95

    Number   Major   Minor   RaidDevice State
       0       8        2        0      spare rebuilding   /dev/sda4
       1       8       18        1      active sync   /dev/sdb4
```

The RAID has now been rebuilt.

#### Adding the label to the SWAP partition (if applicable)

> First, we mount the partition containing our files on `/mnt`. In our example, that partition is `md4`.

Reviewer: ```suggestion
First, we mount the partition containing the root of our operating system on `/mnt`. In our example, that partition is `md4`.
```

```sh
mount /dev/md4 /mnt
```

We add the label back to our swap partition:

```sh
mkswap /dev/sda4 -L swap-sda4
mkswap: /dev/sda4: warning: wiping old swap signature.
Setting up swapspace version 1, size = 512 MiB (536866816 bytes)
LABEL=swap-nvme0n1p4, UUID=b3c9e03a-52f5-4683-81b6-cc10091fcd
```

Next, we mount the following directories to make sure any manipulation we make in the chroot environment works properly:

```sh
mount --types proc /proc /mnt/proc
mount --rbind /sys /mnt/sys
mount --make-rslave /mnt/sys
mount --rbind /dev /mnt/dev
mount --make-rslave /mnt/dev
mount --bind /run /mnt/run
mount --make-slave /mnt/run
```

Next, we access the `chroot` environment:

```sh
chroot /mnt
```

We retrieve the UUIDs of both swap partitions:

```sh
blkid /dev/sda4
blkid /dev/sdb4
```

Example:

```sh

> blkid /dev/sda4

Reviewer: Prompt needed here too
Also, you could just display the UUID to make it easier to read:
```
# blkid -s UUID /dev/sda2
/dev/sda2: UUID="badd9d50-102a-3028-c538-4227eb6ea4d4"
```

/dev/sda4: LABEL="swap-sda4" UUID="b3c9e03a-52f5-4683-81b6-cc10091fcd15" TYPE="swap" PARTLABEL="logical" PARTUUID="cb6022d9-7631-449c-956a-116bcd0e2119"
```

```sh
blkid /dev/sdb4
/dev/sdb4: LABEL="swap-sdb4" UUID="d6af33cf-fc15-4060-a43c-cb3b5537f58a" TYPE="swap" PARTLABEL="logical" PARTUUID="d037c35f-2ddb-40d5-be33-31cc496fe54b"
```

Next, we replace the old UUID of the swap partition (**sda4**) with the new one in `/etc/fstab`:

```sh
nano etc/fstab
```

Example:

```sh
UUID=6abfaa3b-e630-457a-bbe0-e00e5b4b59e5       /       ext4    defaults       0       1
UUID=f925a033-0087-40ec-817e-44efab0351ac       /home   ext4    defaults       0       0
UUID=b7b5dd38-9b51-4282-8f2d-26c65e8d58ec       swap    swap    defaults       0       0
UUID=d6af33cf-fc15-4060-a43c-cb3b5537f58a       swap    swap    defaults       0       0
```

Make sure you replace the proper UUID. In our example above, the UUID to replace is `b7b5dd38-9b51-4282-8f2d-26c65e8d58ec` with the new one `b3c9e03a-52f5-4683-81b6-cc10091fcd15` since the other one belongs to **sdb4**.

Next, we make sure everything is properly mounted:

```sh
 mount -av
/                        : ignored

> mount: (hint) your fstab has been modified, but systemd still uses

Reviewer: Drop these two warning lines, customers do not need to see this, it could be confusing

       the old version; use 'systemctl daemon-reload' to reload.
/boot                    : successfully mounted
/boot/efi                : successfully mounted
swap                     : ignored
swap                     : ignored
```

> We enable the swap partition:

Reviewer: ```suggestion
We activate the swap partition:
```

```sh
swapon -av

swapon: /dev/sda4: found signature [pagesize=4096, signature=swap]
swapon: /dev/sda4: pagesize=4096, swapsize=536870912, devsize=536870912
swapon /dev/sda4
swapon: /dev/sdb4: found signature [pagesize=4096, signature=swap]
swapon: /dev/sdb4: pagesize=4096, swapsize=536870912, devsize=536870912
swapon /dev/sdb4
```

> We exit the Chroot environment with `exit` and unmount all the disks:

Reviewer: ```suggestion
We exit the chroot environment with `exit` and unmount all the disks:
```

```sh

> umount -a

Reviewer: Does that work? Doesn't it attempt to unmount rescue-specific filesystems?
I'd just unmount everything related to the OS on disk:
```suggestion
umount -R /mnt
```

```

> We have now successfully completed the RAID rebuild on the server and we can now reboot the server in normal mode.

Reviewer: ```suggestion
We have now successfully completed the RAID rebuild on the server and we can now reboot it in normal mode.
```

> /// details | **Rebuilding the RAID after the secondary disk is replaced**

Reviewer: What's this?

Author: Hello @sbraz I figured or rather assumed that the process to rebuild the RAID when the main disk fails is slightly different from rebuilding the RAID when a secondary disk fails (in the sense that you do not necessarily need to reboot the server in rescue mode). However, if the section is not necessary, I can get rid of it.

Reviewer: I think I said this in another comment (can't find which one :D): everything is mirrored (except for the ESP for UEFI servers) so as long as we have one healthy disk, no matter which one, the server should boot.

>> sudo sgdisk -G /dev/sdX
>> ```
>> 
>> If you receive a message like this:
>> 
>> ```console
>> Warning: The kernel is still using the old partition table.
>> The new table will be used at the next reboot or after you
>> run partprobe(8) or kpartx(8)
>> The operation has completed successfully.
>> ```
>>
>> You can simply run the `partprobe` command or reboot the server when the raid rebuilt is done.
>>
> **For MBR partitions**
>>
>> ```sh
>> sudo sfdisk -d /dev/sdX | sfdisk /dev/sdX 
>> ```
>>
>> The command should be in this format: `sfdisk -d /dev/healthydisk | sfdisk /dev/newdisk`.
>>

Next, we add the partitions to the RAID:

```sh
sudo mdadm --add /dev/md2 /dev/sdb2

# mdadm: added /dev/sdb2

sudo mdadm --add /dev/md4 /dev/sdb4

# mdadm: re-added /dev/sdb4
```

Use the following command to monitor the RAID rebuild: `cat /proc/mdstat`.

Lastly, we add a label and mount the [SWAP] partition (if applicable):

```sh
sudo  mkswap /dev/sdb4 -L swap-sdb4
```

We retrieve the UUIDs of both swap partitions:

```sh
sudo blkid /dev/sda4
sudo blkid /dev/sdb4
```

We replace the old UUID of the swap partition (**sdb4**) with the new one in `/etc/fstab`:

```sh
sudo nano etc/fstab
```

> Make sure you replace the correct UUID. Reload the system with the following command:

Reviewer: ```suggestion
Make sure you replace the correct UUID.

Then ensure systemd knows about the updated fstab with:
```
